pytest_perf/runner.py

```diff
@@ -74,35 +75,89 @@ class BenchmarkRunner:
+        self.control_env = self._setup_env(upstream_package(baseline, control), *deps)
+        self.experiment_env = self._setup_env(local_package(target), *deps)
+
+    def _setup_env(self, install_context: ContextManager[str], *deps: str):
+        with install_context as package:
+            # Close the install context as soon as the package is installed
+            # to avoid baseline <=> target interference.
+            install_item = f"{package}{self.spec}"
+            target = self.stack.enter_context(pip_run.deps.load(install_item, *deps))
         return pip_run.launch._setup_env(target)
 
-    def run(self, cmd: Command):
-        experiment = self.eval(cmd, env=self.experiment_env)
-        control = self.eval(cmd, env=self.control_env)
-        return Result(control, experiment)
+    def run(self, cmd: Command) -> Result:
+        try:
+            experiment = self.eval(cmd, env=self.experiment_env)
+            control = self.eval(cmd, env=self.control_env)
+            return Result(control, experiment)
+        finally:
+            self.stack.close()
 
-    def eval(self, cmd, **kwargs):
+    def eval(self, cmd: List[str], **kwargs) -> str:
         with tempfile.TemporaryDirectory() as empty:
-            out = subprocess.check_output(cmd, cwd=empty, **_text, **kwargs)
-        val = re.search(r'([0-9.]+ \w+) per loop', out).group(1)
+            out = subprocess.check_output(  # type: ignore
+                cmd, cwd=empty, **_text, **kwargs
+            )
+        val = re.search(r'([0-9.]+ \w+) per loop', out).group(1)  # type: ignore
         return val
 
 
-def upstream_url(extras='', control=None):
+def upstream_url() -> str:
     """
     >>> upstream_url()
-    'pytest-perf@git+https://github.com/jaraco/pytest-perf'
+    'https://github.com/jaraco/pytest-perf'
     """
     cmd = ['git', 'remote', 'get-url', 'origin']
-    origin = subprocess.check_output(cmd, **_text).strip()
-    base, sep, name = origin.rpartition('/')
-    rev = f'@{control}' if control else ''
-    return f'{name}{extras}@git+{origin}{rev}'
+    return subprocess.check_output(cmd, **_text).strip()  # type: ignore
+
+
+@contextlib.contextmanager
+def upstream_package(
+    url: Optional[str] = None, control: Optional[str] = None
+) -> Iterator[str]:
+    """Clone the upstream URL to be used as a local package.
+    This avoids the need of knowing the package name to install optional dependencies.
+    """
+    url = url or upstream_url()
+    rev: List[str] = ['--branch', control] if control else []
+    with tempfile.TemporaryDirectory() as tmp:
+        cmd = ['git', 'clone', '--depth', '1', *rev, url, str(tmp)]
+        subprocess.run(cmd, check=True, **_text)  # type: ignore
+        with local_package(tmp) as target:
+            yield target
+
+            if os.name == "nt":
+                # Avoid cleanup errors on Windows
+                subprocess.run("rmdir /S /Q .git", shell=True, check=False)
```

Comment:
I'm not sure this logic is stable. It's relying on the fact that the shell is `cmd.exe`, which in my opinion has been deprecated since Powershell was released.

```
PS C:\> python
Python 3.10.0 (tags/v3.10.0:b494f59, Oct  4 2021, 19:00:18) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> subprocess.run("rmdir /S /Q .git", shell=True, check=False)
The system cannot find the file specified.
CompletedProcess(args='rmdir /S /Q .git', returncode=2)
```

How does pip do it?